File: 0905-sort-array-by-parity/0905-sort-array-by-parity.py
class Solution:
    def sortArrayByParity(self, nums: List[int]) -> List[int]:
        res = []
        odds, evens = 0, 0

        while odds < len(nums) and nums[odds] % 2 == 0:
            odds += 1

        while evens < len(nums) and nums[evens] % 2 == 1:
            evens += 1

        while evens < len(nums):
            res.append(nums[evens])
            evens += 1
            while evens < len(nums) and nums[evens] % 2 == 1:
                evens += 1

        while odds < len(nums):
            res.append(nums[odds])
            odds += 1
            while odds < len(nums) and nums[odds] % 2 == 0:
                odds += 1

        return res

File: 0905-sort-array-by-parity/0905-sort-array-by-parity.cpp
class Solution {
public:
    vector<int> sortArrayByParity(vector<int>& nums) {
        int l = 0;
        while (l < nums.size()) {
            if (nums[l] % 2 == 0) {
                ++l;
                continue;
            }
            int r = l + 1;
            while (r < nums.size() && nums[r] % 2 == 1) ++r;
            if (r == nums.size()) return nums;
            swap(nums.at(l), nums.at(r));
            ++l;
        }
        return nums;
    }
};